### PR TITLE
numpydoc style(signac/contrib/schema.py)

### PR DESCRIPTION
**Pull request by ac-optimus**
## Description

issue #315 
## Types of Changes

- [x] Documentation update
- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt) and added all related issue and pull request numbers for future reference (if applicable). See example below.

Change of doc string to `numpydoc` in schema.py

### CODE REVIEW

File: signac/contrib/schema.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
-# Copyright (c) 2017 The Regents of the University of Michigan
+# Copyright (c) 2017 The Regents of the University of Michigan.
```
Comment (bdice) [inline signac/contrib/schema.py:1]: Let's revert this change.
```suggestion
# Copyright (c) 2017 The Regents of the University of Michigan
```

File: signac/contrib/schema.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
...
 # All rights reserved.
 # This software is licensed under the BSD 3-Clause License.
+"""Project Schema."""
```
Comment (bdice) [inline signac/contrib/schema.py:4]: I like putting a newline after the module docstring. I think that's a matter of preference.

File: signac/contrib/schema.py [hunk 3]
```diff
@@ -9,19 +10,54 @@
...
+        Job.
+    index :
+        A document index.
+
+    exclude_const : bool
+        Exclude all state point keys that are shared by all jobs within this project.
+
+    Yields
+    ------
+    tupple
```
Comment (bdice) [inline signac/contrib/schema.py:57]:
```suggestion
    tuple
```

File: signac/contrib/schema.py [hunk 4]
```diff
@@ -9,19 +10,54 @@
...
+    """Construct a mapping of types to a set of elements drawn from the input values.
+
+    Parameters
+    ----------
+    values :
+        An iterable of values.
+
+    Returns
+    -------
+    A mapping of types to a set of input values of that type.
```
Comment (bdice) [inline signac/contrib/schema.py:33]:
```suggestion
    defaultdict(set)
        A mapping of types to a set of input values of that type.
```

File: signac/contrib/schema.py [hunk 5]
```diff
@@ -36,6 +72,18 @@ def _build_job_statepoint_index(jobs, exclude_const, index):
...
+
+        Parameters
+        ----------
+        x :
+            dictionary values.
+
+        Returns
+        -------
+        dict
+
```
Comment (bdice) [inline signac/contrib/schema.py:85]:
```suggestion
            Dictionary with ``_DictPlaceholder`` keys removed.
```

File: signac/contrib/schema.py [hunk 6]
```diff
@@ -46,48 +94,93 @@ def remove_dict_placeholder(x):
...
         self._schema = schema
 
     @classmethod
     def detect(cls, statepoint_index):
+        """Detect Project's state point schema.
+
+        Parameters
+        ----------
+        statepoint_index :
+            statepoint index.
```
Comment (bdice) [inline signac/contrib/schema.py:117]:
```suggestion
            State point index.
```

File: signac/contrib/schema.py [hunk 7]
```diff
@@ -46,48 +94,93 @@ def remove_dict_placeholder(x):
...
+        Parameters
+        ----------
+        depth : int
             A non-zero value will return a nested formatting up to the specified depth,
             defaults to 0.
-        :type depth:
-            int
-        :param precision:
+        precision : int
             Round numerical values up the give precision, defaults to unlimited precision.
```
Comment (bdice) [inline signac/contrib/schema.py:136]:
```suggestion
            Round numerical values to the given precision, defaults to unlimited precision.
```

File: signac/contrib/schema.py [hunk 8]
```diff
@@ -46,48 +94,93 @@ def remove_dict_placeholder(x):
...
+
+            Parameters
+            ----------
+            x : int
+                integer value to convert to string.
+
+            Returns
+            -------
+            str
+
```
Comment (bdice) [inline signac/contrib/schema.py:162]:
```suggestion
                Formatted value.
```

File: signac/contrib/schema.py [hunk 9]
```diff
@@ -46,48 +94,93 @@ def remove_dict_placeholder(x):
...
         if max_num_range is None:
             max_num_range = 5
 
         def _fmt_value(x):
+            """Convert x to string.
+
+            Parameters
+            ----------
+            x : int
+                integer value to convert to string.
```
Comment (bdice) [inline signac/contrib/schema.py:157]: This is designed more for `float` than `int`.
```suggestion
            x :
                Value to convert to string.
```

File: signac/contrib/schema.py [hunk 10]
```diff
@@ -46,48 +94,93 @@ def remove_dict_placeholder(x):
...
             A formatted representation of the project schema.
+
         """
         if depth is None:
             depth = 0
         if max_num_range is None:
             max_num_range = 5
 
         def _fmt_value(x):
+            """Convert x to string.
```
Comment (bdice) [inline signac/contrib/schema.py:152]:
```suggestion
            """Convert a value to a string, rounded to a given precision.
```

File: signac/contrib/schema.py [hunk 11]
```diff
@@ -46,48 +94,93 @@ def remove_dict_placeholder(x):
...
+            str
+
+            """
             if precision is not None and isinstance(x, Number):
                 return str(round(x, precision))
             else:
                 return str(x)
 
         def _fmt_range(type_, values):
+            """Convert range of values into a single string.
```
Comment (bdice) [inline signac/contrib/schema.py:170]:
```suggestion
            """Convert sequence of values into a comma-separated string.
            
            Inserts an ellipsis (...) if the number of values exceeds ``max_num_range``.
```

**Comment by bdice** [comment]
@ac-optimus When you are able, please review the comments so far. Please click "Resolve conversation" on all the changes you have applied and only leave the comments that you aren't sure about. Try to fill some of the remaining docstrings, to the best of your ability. Then re-request my review and I can take another pass on anything that's left.

File: signac/contrib/schema.py (hunk 9 above)
Comment (ac-optimus) [reply signac/contrib/schema.py]: I have added type of x as `float and int`. it is cool?

**Comment by ac-optimus** [comment]
> @ac-optimus When you are able, please review the comments so far. Please click "Resolve conversation" on all the changes you have applied and only leave the comments that you aren't sure about. Try to fill some of the remaining docstrings, to the best of your ability. Then re-request my review and I can take another pass on anything that's left.

@bdice  done that.

File: signac/contrib/schema.py [hunk 12]
```diff
@@ -36,6 +74,19 @@ def _build_job_statepoint_index(jobs, exclude_const, index):
     def strip_prefix(key): return k[len('statepoint.'):]
 
     def remove_dict_placeholder(x):
+        """Remove _DictPlaceholder elements from a mapping.
+
+        Parameters
+        ----------
+        x :
+            dictionary values.
```
Comment (bdice) [inline signac/contrib/schema.py:82]:
```suggestion
            Dictionary from which ``_DictPlaceholder`` values will be removed.
```

File: signac/contrib/schema.py [hunk 13]
```diff
@@ -46,48 +97,97 @@ def remove_dict_placeholder(x):
...
+        """Detect Project's state point schema.
+
+        Parameters
+        ----------
+        statepoint_index :
+            State point index.
+
+        Returns
+        -------
+        type : `signac.contrib.schema.ProjectSchema`
```
Comment (bdice) [inline signac/contrib/schema.py:124]:
```suggestion
        ProjectSchema
```

File: signac/contrib/schema.py [hunk 14]
```diff
@@ -46,48 +97,97 @@ def remove_dict_placeholder(x):
...
             depth = 0
         if max_num_range is None:
             max_num_range = 5
 
         def _fmt_value(x):
+            """Convert a value to a string, rounded to a given precision.
+
+            Parameters
+            ----------
+            x : float or int
```
Comment (bdice) [inline signac/contrib/schema.py:159]: The inputs don't have to be numerical, they could be strings. I suggest leaving no type specifier here.
```suggestion
            x :
```

File: signac/contrib/schema.py [hunk 15]
```diff
@@ -46,48 +97,97 @@ def remove_dict_placeholder(x):
...
 
         def _fmt_range(type_, values):
+            """Convert sequence of values into a comma-separated string.
+
+            Inserts an ellipsis (...) if the number of values exceeds ``max_num_range``.
+
+            Parameters
+            ----------
+            type_ :
+                param values.
```
Comment (bdice) [inline signac/contrib/schema.py:181]:
```suggestion
            type_ : type
                Type of values.
```

File: signac/contrib/schema.py [hunk 16]
```diff
@@ -46,48 +97,97 @@ def remove_dict_placeholder(x):
...
+            values :
+                An iterable of values.
+
+            Returns
+            -------
+            str
+                Comma-separated string for squence of values passed.
```
Comment (bdice) [inline signac/contrib/schema.py:188]:
```suggestion
                Comma-separated string of the input values.
```

File: signac/contrib/schema.py [hunk 17]
```diff
@@ -151,35 +263,64 @@ def __iter__(self):
...
+        """Return schema items."""
         return self._schema.items()
 
     def __eq__(self, other):
+        """Check if two schemas are the same.
+
+        Returns
+        -------
+        bool
+            True if both the schema has has the same values.
```
Comment (bdice) [inline signac/contrib/schema.py:283]:
```suggestion
            True if both schemas have the same keys and values.
```

File: signac/contrib/schema.py [hunk 18]
```diff
@@ -151,35 +263,64 @@ def __iter__(self):
...
-        :param ignore_values:
-            Ignore if the value (range) of a specific keys differ, only return missing keys.
-        :type ignore_values:
-            bool
-        :returns:
+        Parameters
+        ----------
+        ignore_values : bool
+            Ignore if the value (range) of a specific keys differ,
+            only return missing keys. (Default value = False)
```
Comment (bdice) [inline signac/contrib/schema.py:295]:
```suggestion
            only return missing keys (Default value = False).
```

File: signac/contrib/schema.py [hunk 19]
```diff
@@ -151,35 +263,64 @@ def __iter__(self):
...
+        """Evaluate the schema for the given state points.
+
+        Parameters
+        ----------
+        jobs_or_statepoints :
+            jobs or statepoints of Project.
+
+        Returns
+        -------
+        type : `signac.contrib.schema.ProjectSchema`
```
Comment (bdice) [inline signac/contrib/schema.py:320]:
```suggestion
        type : ProjectSchema
```

File: signac/contrib/schema.py [hunk 20]
```diff
@@ -151,35 +263,64 @@ def __iter__(self):
...
         return ret
 
     def __call__(self, jobs_or_statepoints):
-        "Evaluate the schema for the given state points."
+        """Evaluate the schema for the given state points.
+
+        Parameters
+        ----------
+        jobs_or_statepoints :
+            jobs or statepoints of Project.
```
Comment (bdice) [inline signac/contrib/schema.py:316]:
```suggestion
            An iterable of jobs or state points.
```

File: signac/contrib/schema.py [hunk 21]
```diff
@@ -9,19 +11,55 @@
...
+        A mapping of types to a set of input values of that type.
+
+    """
     values_by_type = ddict(set)
     for v in values:
         values_by_type[type(v)].add(v)
     return values_by_type
 
 
 def _build_job_statepoint_index(jobs, exclude_const, index):
```
Comment (bdice) [inline signac/contrib/schema.py:44]: The `jobs` argument isn't used here. Since this is a private method, this argument can be removed. I pushed a commit that removed it and updated the docstrings for this method.

Comment (ac-optimus) [reply signac/contrib/schema.py]: thanks!
